src/Umbraco.Infrastructure/Persistence/Repositories/Implement/DistributedJobRepository.cs

```diff
@@ -0,0 +1,124 @@
+﻿using NPoco;
+using Umbraco.Cms.Core.Exceptions;
+using Umbraco.Cms.Infrastructure.BackgroundJobs;
+using Umbraco.Cms.Infrastructure.Models;
+using Umbraco.Cms.Infrastructure.Persistence.Dtos;
+using Umbraco.Cms.Infrastructure.Scoping;
+using Umbraco.Extensions;
+
+namespace Umbraco.Cms.Infrastructure.Persistence.Repositories.Implement;
+
+/// <inheritdoc />
+internal class DistributedJobRepository(IScopeAccessor scopeAccessor) : IDistributedJobRepository
+{
+    /// <inheritdoc />
+    public string? GetRunnable()
+    {
+        if (scopeAccessor.AmbientScope is null)
+        {
+            return null;
+        }
+
+        Sql<ISqlContext> sql = scopeAccessor.AmbientScope.SqlContext.Sql()
+            .Select<DistributedJobDto>()
+            .From<DistributedJobDto>()
+            .Where<DistributedJobDto>(x => x.IsRunning == false)
+            .OrderBy<DistributedJobDto>(x => x.LastRun);
+
+        IUmbracoDatabase database = scopeAccessor.AmbientScope.Database;
+        List<DistributedJobDto> jobs = database.Fetch<DistributedJobDto>(sql);
+
+        // Grab the first runnable job.
+        DistributedJobDto? job = jobs.FirstOrDefault(x => x.LastRun < DateTime.UtcNow - TimeSpan.FromTicks(x.Period));
+
+        if (job is null)
+        {
+            return null;
+        }
+
+        job.LastAttemptedRun = DateTime.UtcNow;
+        job.IsRunning = true;
+        database.Update(job);
+
+
+        return job.Name;
+    }
+
+    /// <inheritdoc />
+    public void Finish(string jobName)
```

Comment:
Same here what it means that a job is finished is domain logic and should be in the service imo

Reply:
Fixed 💪